crates/bevy_gizmos/src/pipeline_3d.rs

```diff
@@ -1,53 +1,97 @@
+        Commands, Query, Res, ResMut, Resource, SystemParamItem,
+    },
     world::{FromWorld, World},
 };
-use bevy_pbr::*;
-use bevy_render::{
-    mesh::Mesh,
-    render_resource::Shader,
-    view::{ExtractedView, ViewTarget},
+use bevy_pbr::{
+    MeshPipeline, MeshPipelineKey, SetMeshViewBindGroup, MAX_CASCADES_PER_LIGHT,
+    MAX_DIRECTIONAL_LIGHTS,
 };
 use bevy_render::{
-    mesh::MeshVertexBufferLayout,
+    extract_component::{ComponentUniforms, DynamicUniformIndex},
     render_asset::RenderAssets,
-    render_phase::{DrawFunctions, RenderPhase, SetItemPipeline},
+    render_phase::{
+        AddRenderCommand, DrawFunctions, PhaseItem, RenderCommand, RenderCommandResult,
+        RenderPhase, SetItemPipeline, TrackedRenderPass,
+    },
     render_resource::*,
+    renderer::RenderDevice,
     texture::BevyDefault,
-    view::Msaa,
+    view::{ExtractedView, Msaa, ViewTarget},
+    Render, RenderApp, RenderSet,
 };
 
-use crate::{GizmoConfig, GizmoMesh, LINE_SHADER_HANDLE};
+pub struct LineGizmo3dPlugin;
+impl Plugin for LineGizmo3dPlugin {
+    fn build(&self, app: &mut App) {
+        let Ok(render_app) = app.get_sub_app_mut(RenderApp) else { return };
 
-#[derive(Resource)]
-pub(crate) struct GizmoPipeline {
+        render_app
+            .add_render_command::<Transparent3d, DrawLineGizmo3d>()
+            .init_resource::<GizmoLinePipeline>()
+            .init_resource::<SpecializedRenderPipelines<GizmoLinePipeline>>()
+            .add_systems(
+                Render,
+                (queue_polyline_bind_group, queue_line_gizmos_3d).in_set(RenderSet::Queue),
+            );
+    }
+}
+
+#[derive(Clone, Resource)]
+struct GizmoLinePipeline {
     mesh_pipeline: MeshPipeline,
-    shader: Handle<Shader>,
+    layout: BindGroupLayout,
 }
 
-impl FromWorld for GizmoPipeline {
+impl FromWorld for GizmoLinePipeline {
     fn from_world(render_world: &mut World) -> Self {
-        GizmoPipeline {
+        let render_device = render_world.resource::<RenderDevice>();
+        let layout = render_device.create_bind_group_layout(&BindGroupLayoutDescriptor {
+            entries: &[BindGroupLayoutEntry {
+                binding: 0,
+                visibility: ShaderStages::VERTEX,
+                ty: BindingType::Buffer {
+                    ty: BufferBindingType::Uniform,
+                    has_dynamic_offset: true,
+                    min_binding_size: Some(LineGizmoUniform::min_size()),
+                },
+                count: None,
+            }],
+            label: Some("polyline_layout"),
+        });
+
+        GizmoLinePipeline {
             mesh_pipeline: render_world.resource::<MeshPipeline>().clone(),
-            shader: LINE_SHADER_HANDLE.typed(),
+            layout,
         }
     }
 }
 
-impl SpecializedMeshPipeline for GizmoPipeline {
-    type Key = (bool, MeshPipelineKey);
+impl SpecializedRenderPipeline for GizmoLinePipeline {
+    type Key = (bool, bool, MeshPipelineKey);
```

Comment:
The tuple could be replaced by a struct with named fields, to clarify the intent.